### ISSUE TITLE
Unify Promises/Error throw in API

### ISSUE THREAD

**Issue by Jesus89** [CartoDB/carto-vl#342]
Now some API functions return Promises, others don't. There are also Errors thrown in between breaking the Promises chain.

We must unify that going for Promises as the first option and using throw for constructors. This will imply changes in the code, tests, documentation and some examples.

I have made a recap of current CARTO VL exceptions. All constructors and global functions can throw Errors and this is OK, so I want to focus on the Layer functions:
- addTo: is an async function but return no Promise, also throws Errors.
- update: is an async function that returns a Promise but also throws Errors.
- blendToViz: is an async function that returns a Promise with the catched Errors.
- _vizChanged: it is not part of the public API but mixes Promises and Errors.

My proposal is to unify this and make all these functions properly async Promises without throwing Errors.

### setDefaultAuth
- CartoError: authRequired
- CartoError: authObjectRequired
- CartoError: apiKeyRequired
- CartoError: apiKeyStringRequired
- CartoError: nonValidApiKey
- CartoError: usernameRequired
- CartoError: usernameStringRequired
- CartoError: nonValidUsername

### setDefaultConfig
- CartoError: configObjectRequired
- CartoError: serverURLStringRequired

### Interactivity

Constructor
- Error: Invalid layer list, parameter must be an array of carto.Layer objects
- Error: Invalid argument, layer list must not be empty
- Error: Invalid layer, layer must be an instance of carto.Layer
- Error: Invalid argument, all layers must belong to the same map

on/off
- Error: Unrecognized event: ${eventName}. Available events: featureClick,featureClickOut,featureEnter,featureHover,featureLeave

### Layer

Constructor
- CartoError: idRequired
- CartoError: idStringRequired
- CartoError: nonValidId
- CartoError: sourceRequired
- CartoError: nonValidSource
- CartoError: vizRequired
- CartoError: nonValidViz
- CartoError: sharedViz

**addTo (no Promise)**
- CartoError: nonValidMap

**update (Promise + Errors) !**
- CartoError: sourceRequired
- CartoError: nonValidSource
- CartoError: vizRequired
- CartoError: nonValidViz
- CartoError: sharedViz
- Error: Another atomic change was done before this one committed

**blendToViz (Promise)**
- reject: CartoError: vizRequired
- reject: CartoError: nonValidViz
- reject: CartoError: sharedViz
- reject: CartoError: vizRequired
- reject: Error: A source is required before changing the viz
- reject: Error: A source change was made before the metadata was retrieved, therefore, metadata is stale and it cannot be longer consumed

(not public API) _vizChanged (Promise) !
- Error: A source is required before changing the viz
- Error: A source change was made before the metadata was retrieved, therefore, metadata is stale and it cannot be longer consumed

### Viz

Constructor
- CartoError: nonValidDefinition
- CartoError: resolutionNumberRequired
- CartoError: resolutionTooSmall
- CartoError: resolutionTooBig
- CartoError: nonValidExpression[]
- Error: Viz contains a circular dependency
- Errors in jsep parser
- Errors in expressions

### source.Dataset

Constructor
- CartoError: tableNameRequired
- CartoError: tableNameStringRequired
- CartoError: nonValidTableName
- CartoError: authRequired
- CartoError: authObjectRequired
- CartoError: apiKeyRequired
- CartoError: apiKeyStringRequired
- CartoError: nonValidApiKey
- CartoError: usernameRequired
- CartoError: usernameStringRequired
- CartoError: nonValidUsername
- CartoError: configObjectRequired
- CartoError: serverURLStringRequired

### source.SQL

Constructor
- CartoError: queryRequired
- CartoError: queryStringRequired
- CartoError: nonValidQuery
- CartoError: nonValidSQLQuery
- CartoError: authRequired
- CartoError: authObjectRequired
- CartoError: apiKeyRequired
- CartoError: apiKeyStringRequired
- CartoError: nonValidApiKey
- CartoError: usernameRequired
- CartoError: usernameStringRequired
- CartoError: nonValidUsername
- CartoError: configObjectRequired
- CartoError: serverURLStringRequired

### source.GeoJSON

Constructor
- CartoError: dataRequired
- CartoError: dataObjectRequired
- CartoError: nonValidGeoJSONData
- CartoError: multipleFeatureTypes[]
- Error: Unsupported GeoJSON: the property '${propertyName}' has different types in different features.

### expressions

blendTo/eval
- Errors in expressions

**Comment by davidmanzanares** [comment]
You cannot throw on constructors in all cases since constructors are synchronous in nature.

cc: @IagoLast

**Comment by Jesus89** [comment]
Now we are using throw in constructors for things like type checking, etc.

**Comment by davidmanzanares** [comment]
Yes, but you cannot use throw when a column is not present in the source for example. Since that operation requires an HTTP request.

**Comment by Jesus89** [comment]
That's correct. We can use events in that case or check other proposals like the factory concept.

**Comment by IagoLast** [comment]
ummm, I think we must re-evaluate the possibility of using factories to generate async objects

**Comment by davidmanzanares** [comment]
Please, @IagoLast evaluate it and create an issue with it. Evaluate the possibility of removing all public classes then. We should decide this ASAP since other efforts (like docs) are going to be wasted if we change this later.

**Comment by IagoLast** [comment]
@davidmanzanares I have a wip [here](https://github.com/CartoDB/carto-vl/issues/321) but I want to go deeper before starting a discussion.

My developer feelings tell me that factories will be better but I want to do a few POC and research a little more to have a strong theory ;) 

We always could use [rxjs](http://reactivex.io/rxjs/) and have all our problems solved ^^

**Comment by rochoa** [comment]
Clarify.

**Comment by Jesus89** [comment]
Updated description with a recap of CARTO VL exceptions and the conflictive functions to be fixed.

**Comment by davidmanzanares** [comment]
Regarding `addTo`, I don't see exactly where it can throw asynchronously, however, fixing this may be redundant with adapting the integration with the upstream custom layers branch.

Regarding `update`, that is an `async` function, the throws I see there will reject the promise and are from my point of view, correct. If not, please, clarify.

Regarding `blendToViz`, that function has a problem because it uses a try catch to then return a new rejected promise. Since this is an async function, it can throw errors and they will be automatically transformed into the rejection of the promise.

> _vizChanged: it is not part of the public API but mixes Promises and Errors.

Async functions can be mixed with throws without problems, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

**Comment by Jesus89** [comment]
Yes, the point is to have a unified structure. Are you OK if we remove the `try catch` from the `blendToViz` to have the same structure than `update`?

**Comment by davidmanzanares** [comment]
Of course, that `try catch` is not needed

**Comment by Jesus89** [comment]
A bit off topic but, what do you think about `addTo` function?

This function is part of our current API to avoid using always `map.on('load'` as MapboxGL does. It tries to add the layer and if this fails with a `Style is not done loading` error it subscribes to `map.on('load'` with the passed callback.

With the new MGL integration our `carto.Layer` will be a native MGL layer so we don't need a specific function `addTo` and we can use directly their API instead:

```js
map.on('load', () => {
  map.addLayer(cartoLayer);
});
```

The question is what happens with the `addTo` function. Are we going to change it, are we going to keep it?

**Comment by davidmanzanares** [comment]
I think that we should better discuss that when we work on the integration with their new functionality.

**Comment by rochoa** [comment]
We might want to keep it, but we could do:

```js
layer.addTo = map => {
  map.addLayer(this);
}
```

**Comment by Jesus89** [comment]
Yes, we will discuss more in the integration thread.

I like the @rochoa's approach but we should keep in mind that `map.on('load'` will be always with us.
